File: BasicFullscreenHandling/src/main/java/com/bitmovin/player/samples/fullscreen/basic/CustomFullscreenHandler.java
/*
 * Bitmovin Player Android SDK
 * Copyright (C) 2017, Bitmovin GmbH, All Rights Reserved
 *
 * This source code and its use and distribution, is subject to the terms
 * and conditions of the applicable license agreement.
 */

package com.bitmovin.player.samples.fullscreen.basic;


import android.annotation.SuppressLint;
import android.app.Activity;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import android.view.View;
import android.view.ViewGroup;

import com.bitmovin.player.PlayerView;
import com.bitmovin.player.api.ui.FullscreenHandler;

public class CustomFullscreenHandler implements FullscreenHandler {
    private Activity activity;
    private View decorView;
    private PlayerView playerView;
    private ActionBar actionBar;

    private PlayerOrientationListener playerOrientationListener;

    private boolean isFullscreen;

    public CustomFullscreenHandler(Activity activity,
                                   PlayerView playerView,
                                   @Nullable
                                   ActionBar actionbar) {
        this.activity = activity;
        this.playerView = playerView;
        this.actionBar = actionbar;
        decorView = activity.getWindow().getDecorView();
        playerOrientationListener = new PlayerOrientationListener(activity);

        playerOrientationListener.enable();

        ViewCompat.setOnApplyWindowInsetsListener(activity.findViewById(R.id.root), (view, insets) -> {
            if (isFullscreen) {
                view.setPadding(0, 0, 0, 0);
            } else {
                Insets bars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
                view.setPadding(bars.left, bars.top, bars.right, bars.bottom);
            }
            return WindowInsetsCompat.CONSUMED;
        });
    }

    private void handleFullscreen(boolean fullscreen) {
        this.isFullscreen = fullscreen;

        doSystemUiVisibility(fullscreen);
        doLayoutChanges(fullscreen);
    }

    private void doSystemUiVisibility(final boolean fullScreen) {
        this.decorView.post(() -> {
            int uiParams = getSystemUiVisibilityFlags(fullScreen);

            decorView.setSystemUiVisibility(uiParams);
        });
    }

    private void doLayoutChanges(final boolean fullscreen) {
        Looper mainLooper = Looper.getMainLooper();
        boolean isAlreadyMainLooper = Looper.myLooper() == mainLooper;

        UpdateLayoutRunnable updateLayoutRunnable = new UpdateLayoutRunnable((AppCompatActivity) activity, fullscreen);

        if (isAlreadyMainLooper) {
            updateLayoutRunnable.run();
        } else {
            Handler handler = new Handler(mainLooper);
            handler.post(updateLayoutRunnable);
        }
    }

    @Override
    public void onFullscreenRequested() {
        handleFullscreen(true);
    }

    @Override
    public void onFullscreenExitRequested() {
        handleFullscreen(false);
    }

    @Override
    public void onResume() {
        if (isFullscreen) {
            doSystemUiVisibility(isFullscreen);
        }
    }

    @Override
    public void onPause() {}

    @Override
    public void onDestroy() {
        playerOrientationListener.disable();
    }

    @Override
    public boolean isFullscreen() {
        return isFullscreen;
    }

    private class UpdateLayoutRunnable implements Runnable {
        private AppCompatActivity activity;
        private boolean fullscreen;

        private UpdateLayoutRunnable(AppCompatActivity activity, boolean fullscreen) {
            this.activity = activity;
            this.fullscreen = fullscreen;
        }

        @Override
        @SuppressLint("RestrictedApi")
        public void run() {
            if (actionBar != null) {
                if (this.fullscreen) {
                    actionBar.hide();
                } else {
                    actionBar.show();
                }
            }

            if (CustomFullscreenHandler.this.playerView.getParent() instanceof ViewGroup) {
                ViewGroup parentView = (ViewGroup) CustomFullscreenHandler.this.playerView.getParent();

                for (int i = 0; i < parentView.getChildCount(); i++) {
                    View child = parentView.getChildAt(i);

                    if (child != playerView) {
                        child.setVisibility(fullscreen ? View.GONE : View.VISIBLE);
                    }
                }
            }
        }
    }

    public int getSystemUiVisibilityFlags(boolean fullScreen) {
        if (fullScreen) {
            return View.SYSTEM_UI_FLAG_HIDE_NAVIGATION |
                    View.SYSTEM_UI_FLAG_FULLSCREEN |
                    View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION |
                    View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN |
                    View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY;
        } else {
            return View.SYSTEM_UI_FLAG_VISIBLE;
        }
    }
}

File: BasicFullscreenHandlingKotlin/src/main/java/com/bitmovin/player/samples/fullscreen/basic/CustomFullscreenHandler.kt
package com.bitmovin.player.samples.fullscreen.basic

import android.app.Activity
import android.os.Handler
import android.os.Looper
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.ActionBar
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updatePadding
import com.bitmovin.player.PlayerView
import com.bitmovin.player.api.ui.FullscreenHandler

class CustomFullscreenHandler(
    activity: Activity,
    root:View,
    private val playerView: PlayerView,
    private val actionbar: ActionBar?
) : FullscreenHandler {
    override var isFullscreen = false
    private var decorView: View? = activity.window.decorView
    private val playerOrientationListener = PlayerOrientationListener(activity).apply { enable() }

    init {
        ViewCompat.setOnApplyWindowInsetsListener(root) { view, windowInsets ->
            if (isFullscreen) {
                view.updatePadding(0, 0, 0, 0)
            } else {
                val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())
                view.updatePadding(insets.left, insets.top, insets.right, insets.bottom)
            }
            WindowInsetsCompat.CONSUMED
        }
    }

    private fun handleFullscreen(fullscreen: Boolean) {
        isFullscreen = fullscreen
        doSystemUiVisibility(fullscreen)
        doLayoutChanges()
    }

    private fun doSystemUiVisibility(fullScreen: Boolean) {
        decorView?.post {
            val uiParams = getSystemUiVisibilityFlags(fullScreen)
            decorView?.systemUiVisibility = uiParams
        }
    }

    private fun doLayoutChanges() {
        val mainLooper = Looper.getMainLooper()
        val isAlreadyMainLooper = Looper.myLooper() == mainLooper

        if (isAlreadyMainLooper) {
            updateLayout()
        } else {
            val handler = Handler(mainLooper)
            handler.post(::updateLayout)
        }
    }

    private fun updateLayout() {
        val parentView = playerView.parent
        if (isFullscreen) {
            actionbar?.hide()
        } else {
            actionbar?.show()
        }

        if (parentView !is ViewGroup) return

        for (i in 0 until parentView.childCount) {
            parentView
                .getChildAt(i)
                .takeIf { it !== playerView }
                ?.visibility = if (this.isFullscreen) View.GONE else View.VISIBLE
        }
    }

    override fun onFullscreenRequested() = handleFullscreen(true)

    override fun onFullscreenExitRequested() = handleFullscreen(false)

    override fun onResume() {
        if (isFullscreen) {
            doSystemUiVisibility(isFullscreen)
        }
    }

    override fun onPause() {}

    override fun onDestroy() = playerOrientationListener.disable()
}

fun getSystemUiVisibilityFlags(fullScreen: Boolean): Int = if (!fullScreen) {
    View.SYSTEM_UI_FLAG_VISIBLE
} else {
    View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or
            View.SYSTEM_UI_FLAG_FULLSCREEN or
            View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION or
            View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or
            View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
}